From 39c94f63a62c2991cce1b5d577da35545203ac3d Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Tue, 16 May 2023 10:47:33 +0800
Subject: [PATCH 018/263] [Huawei] userswap: check read and write permissions
 for swap-out pages

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

Check the VM_READ and VM_WRITE flags of vma->vm_flags to determine
whether the read and write permissions of the swap-out page VA are
consistent with those of the swap-out buffer VA. If they are
inconsistent, the swap operation will fail.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index c9ac7e6389a7..dcc608e0d8fe 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -311,8 +311,8 @@ static unsigned long do_user_swap(struct mm_struct *mm,
 			goto out_recover;
 
 		ret = -EACCES;
-		if (pgprot_val(old_vma->vm_page_prot) !=
-		    pgprot_val(new_vma->vm_page_prot))
+		if (!(old_vma->vm_flags & VM_WRITE) &&
+		    (new_vma->vm_flags & VM_WRITE))
 			goto out_recover;
 
 		ret = -ENXIO;
-- 
2.17.1

